pkg/controllers/managementuser/snapshotbackpopulate/snapshotbackpopulate.go

```diff
@@ -77,58 +90,195 @@ func (h *handler) OnChange(key string, configMap *corev1.ConfigMap) (*corev1.Con
+	ls, err := labels.Parse(fmt.Sprintf("%s=%s", capi.ClusterLabelName, cluster.Name))
+	if err != nil {
+		return false, nil, err
+	}
+	machines, err := h.machineCache.List(cluster.Namespace, ls)
+	if err != nil {
+		return false, nil, err
+	}
+	for _, machine := range machines {
+		if machine.Status.NodeRef != nil && machine.Status.NodeRef.Name == nodeName {
+			return true, machine, nil
+		}
+	}
+	return true, nil, fmt.Errorf("unable to find node %s in machines", nodeName)
+}
+
+func (h *handler) configMapToSnapshots(configMap *corev1.ConfigMap, cluster *provv1.Cluster) (result map[string]rkev1.ETCDSnapshot, _ error) {
+	clusterName := cluster.Name
+	clusterNamespace := cluster.Namespace
+	result = map[string]rkev1.ETCDSnapshot{}
 	for k, v := range configMap.Data {
 		file := &snapshotFile{}
 		if err := json.Unmarshal([]byte(v), file); err != nil {
 			logrus.Errorf("invalid non-json value in %s/%s for key %s in cluster %s", configMap.Namespace, configMap.Name, k, h.clusterName)
 			return nil, nil
 		}
+		snapshotName := clusterName + "-" + file.Name
+		// Validate that the corresponding machine for the node exists before creating the snapshot
 		snapshot := rkev1.ETCDSnapshot{
-			Name:      file.Name,
-			Location:  file.Location,
-			Metadata:  file.Metadata,
-			Message:   file.Message,
-			NodeName:  file.NodeName,
-			CreatedAt: file.CreatedAt,
-			Size:      file.Size,
-			Status:    file.Status,
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      snapshotName,
+				Namespace: clusterNamespace,
+				Labels: map[string]string{
+					rke2.ClusterNameLabel: clusterName,
+					rke2.NodeNameLabel:    file.NodeName,
+				},
+				OwnerReferences: []metav1.OwnerReference{},
+			},
+			SnapshotFile: rkev1.ETCDSnapshotFile{
+				Name:      file.Name,
+				Location:  file.Location,
+				Metadata:  file.Metadata,
+				Message:   file.Message,
+				NodeName:  file.NodeName,
+				CreatedAt: file.CreatedAt,
+				Size:      file.Size,
+				Status:    file.Status,
+			},
 		}
+		fileSuffix := "-local"
 		if file.S3 != nil {
-			snapshot.S3 = &rkev1.ETCDSnapshotS3{
+			fileSuffix = "-s3"
+			snapshot.SnapshotFile.S3 = &rkev1.ETCDSnapshotS3{
 				Endpoint:      file.S3.Endpoint,
 				EndpointCA:    file.S3.EndpointCA,
 				SkipSSLVerify: file.S3.SkipSSLVerify,
 				Bucket:        file.S3.Bucket,
 				Region:        file.S3.Region,
 				Folder:        file.S3.Folder,
 			}
+			snapshot.OwnerReferences = append(snapshot.OwnerReferences, metav1.OwnerReference{
+				APIVersion:         cluster.APIVersion,
+				Kind:               cluster.Kind,
+				Name:               cluster.Name,
+				UID:                cluster.UID,
+				Controller:         &[]bool{true}[0],
+				BlockOwnerDeletion: &[]bool{true}[0],
+			})
+		} else {
+			listSuccessful, machine, err := h.retrieveMachineFromNode(file.NodeName, cluster)
+			if listSuccessful && err != nil {
+				continue // don't add this snapshot to the list as we can't actually correlate it to an existing node
```

Comment:
No, because there could be a bunch of reasons this is failing and I don't think we need to log anything for it. I will eat my own words if that turns out to not to be the case though